qiskit/compiler/assembler.py

```diff
@@ -293,7 +293,7 @@ def _parse_common_args(
     max_shots = getattr(backend_config, "max_shots", None)
     if shots is None:
         if max_shots:
-            shots = min(1024, max_shots)
+            shots = min(backend._options.shots, max_shots)
```

Comment:
Looking at this issue in more detail, it seems like the `shots` option is only defined on https://github.com/Qiskit/qiskit-ibmq-provider/blob/4f6b7f693d0b6f8afd63f389bb7576ac1be70b91/qiskit/providers/ibmq/ibmqbackend.py#L136 and therefore this change should not be made to `assemble` which should work for any backend (including non-IBM backends). 




It may be that the original issue, should not be addressed as the `assemble` path is to be deprecated long-term. `self._options.shots` should be used in assemble by the backend here - https://github.com/Qiskit/qiskit-ibmq-provider/blob/4f6b7f693d0b6f8afd63f389bb7576ac1be70b91/qiskit/providers/ibmq/ibmqbackend.py#L306. What do you think @mtreinish @jyu00?

Reply:
Yeah, I'm on the fence here too. One hand backend options are supposed to be a level above assemble. Moving forward `assemble()` is really just going to be an internal implementation detail for the ibm provider and not something users should need or want to use. The provider when generating a qobj payload it should just call it with the appropriate runtime arguments so the generated qobj uses the current set value (either the state in the option or a kwarg to run). But on the other hand using the set option value in the absence of a kwarg does seem to be the more modern way to interact with backends and `assemble()` being a function in terra does make me feel like we should respect that setting here if it exists. So I guess I'm fine either way as I don't have a super strong opinion as ideally I'd like to see us move away from maintaining `assemble()` in terra.

That being said if we do change the behavior here you should not access the inner `_options` attribute like this but instead go through the user facing api `getattr(backend.options, 'shots', 0)` especially since it's not guaranteed a backend has `shots` set (although in practice it's probably the only universal option for all backends). Also, you probably should update the logic here to only put shots in the output when a backend is set if `hasattr(backend.options, 'shots')` is True.